Add InputSource and InputFormat interfaces

jihoonson opened the pull request:
### Description

This is the First PR for https://github.com/apache/incubator-druid/issues/8812 which includes the new interfaces proposed in #8812. A couple of implementations are also included such as `LocalInputSource` and `HttpInputSource` for `InputSource`, and `CsvInputFormat` and `JsonInputFormat` for `InputFormat`. Their formats are:

```json
"inputSource": {
  "type" : "local",
  "baseDir" : "/path/to/dir",
  "filter" : "your filter"
}
```

```json
"inputSource": {
  "type" : "http",
  "uris" : ["http://example.com/uri1", "http://example.com/uri2"],
  "httpAuthenticationUsername": "username",
  "httpAuthenticationPassword": "password provider"
}
```

```json
"inputFormat": {
  "type": "csv",
  "columns": [ "col1", "col2", "col3" ],
  "listDelimiter": "|",
  "findColumnsFromHeader" : true,
  "skipHeaderRows" : 3
}
```

```json
"inputFormat": {
  "type": "json",
  "flattenSpec": {
    // your flatten spec
  }
}
```

Note that both `inputSource` and `inputFormat` are in `ioConfig` as below:

```json
"ioConfig": {
  "type" : "index" or "index_parallel",
  "inputSource" : {
    // your input source
  },
  "inputFormat": {
    // your input format
  }
}
```

These are supported only by native batch indexing tasks yet. Sampler doesn't support them yet.

The old `firehose` and `parser` parameters should still work, but you cannot mix them. Only the combinations of `firehose` + `parser` or `inputSource`  + `inputFormat` are allowed.

Documents will be added after more inputSources and inputFormats are implemented in follow-up PRs.

<hr>

This PR has:
- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/incubator-druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/incubator-druid/blob/master/licenses.yaml)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

<!-- Check the items by putting "x" in the brackets for the done things. Not all of these items apply to every PR. Remove the items which are not done or not relevant to the PR. None of the items from the checklist above are strictly necessary, but it would be very helpful if you at least self-review the PR. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/apache/incubator-druid/8823)
<!-- Reviewable:end -->

vogievetsky commented:
This is amazing.

fjy commented:
One of the most exciting PRs on Druid ingestion in awhile. Glad we got it out.

ccaominh commented on indexing-service/src/main/java/org/apache/druid/indexing/common/task/IndexTask.java:
```diff
@@ -1043,30 +1082,90 @@ public IndexTuningConfig getTuningConfig()
+          ImmutableList.of(new Property<>("firehose", firehoseFactory), new Property<>("inputSource", inputSource))
+      );
+      if (firehoseFactory != null && inputFormat != null) {
+        throw new IAE("Cannot use firehose and inputFormat together. Try use inputSource instead of firehose.");
```
Typo: Try use inputFormat -> Try using inputSource

jihoonson replied:
Fixed.

ccaominh commented on indexing-service/src/main/java/org/apache/druid/indexing/common/task/batch/parallel/ParallelIndexIngestionSpec.java:
```diff
@@ -39,6 +44,18 @@ public ParallelIndexIngestionSpec(
+    );
+    if (dataSchema.getParserMap() != null && ioConfig.getInputSource() != null) {
+      if (!(ioConfig.getInputSource() instanceof FirehoseFactoryToInputSourceAdaptor)) {
+        throw new IAE("Cannot use parser and inputSource together. Try use inputFormat instead of parser.");
```
Typo: Try use inputFormat -> Try using inputFormat

jihoonson replied:
Fixed.

ccaominh commented on indexing-service/src/test/java/org/apache/druid/indexing/common/task/IndexTaskTest.java:
```diff
@@ -115,47 +118,49 @@
+        new Object[]{LockGranularity.TIME_CHUNK, false},
+        new Object[]{LockGranularity.TIME_CHUNK, true},
+        new Object[]{LockGranularity.SEGMENT, false},
+        new Object[]{LockGranularity.SEGMENT, true}
```
This is a relatively slow test (~15 seconds per parameterized run), so all the permutations may be overkill. Perhaps remove (SEGMENT, false), which will still give coverage of both lock granularities and both with/without the input format API.

jihoonson replied:
Removed.

ccaominh commented on indexing-service/src/test/java/org/apache/druid/indexing/common/task/batch/parallel/MultiPhaseParallelIndexingTest.java:
```diff
@@ -90,24 +91,29 @@
-        new Object[]{LockGranularity.SEGMENT}
+        new Object[]{LockGranularity.TIME_CHUNK, false},
+        new Object[]{LockGranularity.TIME_CHUNK, true},
+        new Object[]{LockGranularity.SEGMENT, false},
```
Similar comment to `IndexingTest` about skipping this permutation

ccaominh commented on indexing-service/src/test/java/org/apache/druid/indexing/common/task/batch/parallel/SinglePhaseParallelIndexingTest.java:
```diff
@@ -67,21 +68,26 @@
-        new Object[]{LockGranularity.SEGMENT}
+        new Object[]{LockGranularity.TIME_CHUNK, false},
+        new Object[]{LockGranularity.TIME_CHUNK, true},
+        new Object[]{LockGranularity.SEGMENT, false},
```
Similar comment to `IndexingTest` about skipping this permutation

jihoonson commented on core/src/main/java/org/apache/druid/data/input/FirehoseFactoryToInputSourceAdaptor.java:
```diff
@@ -0,0 +1,96 @@
+    if (firehoseFactory.isSplittable()) {
+      return ((FiniteFirehoseFactory) firehoseFactory).getSplits(splitHintSpec);
+    } else {
+      throw new UnsupportedOperationException();
```
No, only splittable firehose can create splits.

jihoonson commented on core/src/test/java/org/apache/druid/data/input/FirehoseFactoryToInputSourceAdaptorTest.java:
```diff
@@ -0,0 +1,177 @@
+    }
+  }
+
+  private static class TestCsvParseSpec extends CSVParseSpec
```
👍 renamed.

jihoonson commented on indexing-hadoop/src/main/java/org/apache/druid/indexer/HadoopDruidIndexerConfig.java:
```diff
@@ -579,8 +579,9 @@ public void intoConfiguration(Job job)
+    Preconditions.checkNotNull(schema.getDataSchema().getParser(), "inputRowParser");
     Preconditions.checkNotNull(schema.getDataSchema().getParser().getParseSpec(), "parseSpec");
-    Preconditions.checkNotNull(schema.getDataSchema().getParser().getParseSpec().getTimestampSpec(), "timestampSpec");
+    Preconditions.checkNotNull(schema.getDataSchema().getNonNullTimestampSpec(), "timestampSpec");
```
Removed.

jihoonson commented on indexing-service/src/main/java/org/apache/druid/indexing/common/TaskToolbox.java:
```diff
@@ -294,7 +294,7 @@ public IndexMergerV9 getIndexMergerV9()
-  public File getFirehoseTemporaryDir()
+  public File getIndexingTmpDir()
   {
     return new File(taskWorkDir, "firehose");
```
Renamed to `indexing-tmp`.

jihoonson commented on server/src/main/java/org/apache/druid/segment/indexing/DataSchema.java:
```diff
@@ -53,38 +57,53 @@
       @JsonProperty("transformSpec") TransformSpec transformSpec,
-      @JacksonInject ObjectMapper jsonMapper
+      @Deprecated @JsonProperty("parser") @Nullable Map<String, Object> parserMap,
+      @JacksonInject ObjectMapper objectMapper
```
I would rather not adding it because it's never null in production code.

jihoonson commented on server/src/main/java/org/apache/druid/segment/indexing/DataSchema.java:
```diff
@@ -53,38 +57,53 @@
+    this.dimensionsSpec = dimensionsSpec == null
+                          ? null
+                          : computeDimensionsSpec(
+                              Preconditions.checkNotNull(timestampSpec, "timestampSpec"),
```
The original null check has no point since `timestampSpec` is never null. See the constructor of `parseSpec`.

jihoonson commented on server/src/main/java/org/apache/druid/segment/indexing/DataSchema.java:
```diff
@@ -115,83 +158,75 @@ static void validateDatasourceName(String dataSource)
-  public Map<String, Object> getParserMap()
+  @Nullable
+  @JsonProperty
+  public TimestampSpec getTimestampSpec()
```
👍

jihoonson commented on server/src/main/java/org/apache/druid/segment/indexing/DataSchema.java:
```diff
@@ -212,25 +247,78 @@ public TransformSpec getTransformSpec()
            ", aggregators=" + Arrays.toString(aggregators) +
            ", granularitySpec=" + granularitySpec +
            ", transformSpec=" + transformSpec +
+           (parserMap == null ? "" : ", parserMap=" + parserMap) +
```
Changed.

jihoonson commented on core/src/main/java/org/apache/druid/data/input/impl/JSONParseSpec.java:
```diff
@@ -73,6 +74,13 @@ public void verify(List<String> usedCols)
     return new JSONPathParser(getFlattenSpec(), objectMapper);
   }
 
+  @Nullable
```
Removed.

jihoonson commented on core/src/main/java/org/apache/druid/data/input/SplitSource.java:
```diff
@@ -0,0 +1,111 @@
+{
+  Logger LOG = new Logger(SplitSource.class);
+
+  int DEFAULT_FETCH_BUFFER_SIZE = 4 * 1024; // 4 KB
```
This will be used in follow-up PRs.

jihoonson commented on core/src/main/java/org/apache/druid/data/input/impl/FileSource.java:
```diff
@@ -0,0 +1,84 @@
+  FileSource(InputSplit<File> split) throws FileNotFoundException
+  {
+    this.split = split;
+    final RandomAccessFile file = new RandomAccessFile(split.get(), "r");
```
Oh, I really didn't close resources in many places. Thanks.

jihoonson replied:
Hmm this is still not fixed correctly. Will fix it soon

jihoonson replied:
Removed `randomAccessFile` since it's not much useful for now. We can add it back later when we can really benefit from using NIO.

jihoonson commented on core/src/main/java/org/apache/druid/data/input/impl/FirehoseToInputSourceReaderAdaptor.java:
```diff
@@ -0,0 +1,122 @@
+          return firehose.hasMore();
+        }
+        catch (IOException e) {
+          throw new RuntimeException(e);
```
I changed but what's benefit of using it?

ccaominh replied:
I believe it was created for use cases like conforming to the Iterator interface.

jihoonson commented on core/src/main/java/org/apache/druid/data/input/impl/CsvInputFormat.java:
```diff
@@ -0,0 +1,144 @@
+      @JsonProperty("columns") @Nullable List<String> columns,
+      @JsonProperty("listDelimiter") String listDelimiter,
+      @Deprecated @JsonProperty("hasHeaderRow") @Nullable Boolean hasHeaderRow,
+      @JsonProperty("findColumnsFromHeader") @Nullable Boolean findColumnsFromHeader,
```
It can be null if `hasHeaderRow` is specified.

jihoonson commented on core/src/main/java/org/apache/druid/data/input/impl/CsvReader.java:
```diff
@@ -0,0 +1,134 @@
+  @Override
+  public int getNumHeaderLines()
+  {
+    return (hasHeaderRow ? 1 : 0) + skipHeaderRows;
```
Oops, fixed.

jihoonson commented on core/src/test/java/org/apache/druid/data/input/impl/CsvReaderTest.java:
```diff
@@ -0,0 +1,178 @@
+  }
+
+  @Test
+  public void testFindColumnAndSkipHeaders() throws IOException
```
Fixed.

jihoonson commented on core/src/main/java/org/apache/druid/data/input/SplitSource.java:
```diff
@@ -0,0 +1,111 @@
+    final File tempFile = File.createTempFile("druid-split", ".tmp", temporaryDirectory);
+    LOG.debug("Fetching split into file[%s]", tempFile.getAbsolutePath());
+    FileUtils.copyLarge(
+        open(),
```
Oops, yeah it must be closed. Thanks.

jihoonson commented on core/src/main/java/org/apache/druid/data/input/impl/JsonInputFormat.java:
```diff
@@ -0,0 +1,94 @@
+
+  @JsonCreator
+  public JsonInputFormat(
+      @JsonProperty("flattenSpec") JSONPathSpec flattenSpec,
```
Added.

jihoonson commented on core/src/main/java/org/apache/druid/data/input/impl/JsonInputFormat.java:
```diff
@@ -0,0 +1,94 @@
+  @JsonCreator
+  public JsonInputFormat(
+      @JsonProperty("flattenSpec") JSONPathSpec flattenSpec,
+      @JsonProperty("featureSpec") @Nullable Map<String, Boolean> featureSpec
```
Yes, it should be added.

jihoonson commented on core/src/main/java/org/apache/druid/data/input/impl/LocalInputSource.java:
```diff
@@ -0,0 +1,150 @@
+    while (fileIterator.hasNext()) {
+      fileIterator.next();
+      num++;
+    }
```
Changed.

jihoonson commented on core/src/main/java/org/apache/druid/data/input/impl/SplitIteratingReader.java:
```diff
@@ -0,0 +1,138 @@
+        return rowIterator.next();
+      }
+
+      private void checkRowIterator()
```
Changed to `updateRowIteratorIfNeeded()`.

jihoonson commented on core/src/test/java/org/apache/druid/data/input/impl/SplitIteratingReaderTest.java:
```diff
@@ -0,0 +1,98 @@
+        Assert.assertEquals(Integer.toString(i + 1), Iterables.getOnlyElement(row.getDimension("score")));
+        i++;
+      }
+    }
```
👍

jihoonson commented on core/src/test/java/org/apache/druid/data/input/impl/NoopInputSource.java:
```diff
@@ -0,0 +1,46 @@
+
+public class NoopInputSource implements InputSource
+{
+  @Override
```
Changed to throw `UnsupportedOperationException`.

jihoonson commented on core/src/test/java/org/apache/druid/data/input/impl/NoopFirehoseFactory.java:
```diff
@@ -0,0 +1,57 @@
+    return "NoopFirehoseFactory{}";
+  }
+
+  @Override
```
Changed to throw `UnsupportedOperationException`.

jihoonson commented on core/src/test/java/org/apache/druid/data/input/impl/NoopFirehoseFactory.java:
```diff
@@ -0,0 +1,57 @@
+    return 0;
+  }
+
+  @Override
```
Changed to throw `UnsupportedOperationException`.

jihoonson commented on core/src/test/java/org/apache/druid/data/input/impl/NoopFirehoseFactory.java:
```diff
@@ -0,0 +1,57 @@
+  }
+
+  @Override
+  public int getNumSplits(@Nullable SplitHintSpec splitHintSpec) throws IOException
```
Removed.

jihoonson commented on core/src/test/java/org/apache/druid/data/input/impl/JsonReaderTest.java:
```diff
@@ -0,0 +1,116 @@
+        Assert.assertTrue(row.getDimension("path_omg2").isEmpty());
+        Assert.assertTrue(row.getDimension("jq_omg2").isEmpty());
+      }
+    }
```
Added.

jihoonson commented on core/src/test/java/org/apache/druid/data/input/impl/JsonReaderTest.java:
```diff
@@ -0,0 +1,116 @@
+        Assert.assertEquals(Collections.emptyList(), row.getDimension("foo"));
+        Assert.assertTrue(row.getDimension("baz").isEmpty());
+      }
+    }
```
Added.

jihoonson commented on core/src/test/java/org/apache/druid/data/input/impl/CsvInputFormatTest.java:
```diff
@@ -0,0 +1,70 @@
+  }
+
+  @Test
+  public void testComma()
```
Added a new test because `testColumnMissing()` will be removed in https://github.com/apache/incubator-druid/pull/8744.

jihoonson commented on core/src/main/java/org/apache/druid/java/util/common/FileUtils.java:
```diff
@@ -306,6 +306,33 @@ public void close()
     }
   }
 
+  public static long copyLarge(
```
Refactored. I think it's better to add unit tests later to make not this PR bigger.

jihoonson replied:
I was wrong and reverted this change. I think the existing method will be removed together when old firehoseFactory will be gone and want to leave it for now. Maybe I will refactor in a follow-up pr.

jihoonson commented on indexing-service/src/main/java/org/apache/druid/indexing/common/task/IndexTask.java:
```diff
@@ -1015,6 +1044,16 @@ public IndexIngestionSpec(
+          )
+      );
+      if (dataSchema.getParserMap() != null && ioConfig.getInputSource() != null) {
+        throw new IAE("Cannot use parser and inputSource together. Try use inputFormat instead of parser.");
```
Fixed.

jihoonson commented:
I'll fix `CsvReader.verify()` properly once https://github.com/apache/incubator-druid/pull/8744 gets merged.

jon-wei commented on core/src/main/java/org/apache/druid/data/input/ObjectSource.java:
```diff
@@ -0,0 +1,116 @@
+  }
+
+  /**
+   * {@link #fetch} will retry during the fetch if it sees an exception mathing to the returned predicate.
```
mathing -> matching

jihoonson replied:
Thanks, fixed.

jon-wei commented on core/src/main/java/org/apache/druid/data/input/ObjectSource.java:
```diff
@@ -0,0 +1,116 @@
+  /**
+   * {@link #fetch} will retry during the fetch if it sees an exception mathing to the returned predicate.
+   */
+  Predicate<Throwable> getRetryCondition();
```
Suggest calling this `getFetchRetryCondition()` since it's only used for fetch

jihoonson replied:
👍

ccaominh commented on core/src/main/java/org/apache/druid/data/input/ObjectSource.java:
```diff
@@ -50,41 +50,46 @@
+      );
+    }
+    catch (IOException e) {
+      throw new RuntimeException(e);
```
Is converting to a `RuntimeException` desired? The method declares `IOException` as a thrown exception.

jihoonson replied:
Removed.

ccaominh commented on core/src/main/java/org/apache/druid/data/input/Formattable.java:
```diff
@@ -0,0 +1,39 @@
+import javax.annotation.Nullable;
+import java.io.File;
+
+public interface Formattable
```
Do you want to add javadocs (i.e., what does it mean to be "formattable")?

jihoonson replied:
Looks like you're looking at a stale commit. These are removed.

ccaominh commented on core/src/main/java/org/apache/druid/data/input/Unformattable.java:
```diff
@@ -0,0 +1,40 @@
+import javax.annotation.Nullable;
+import java.io.File;
+
+public interface Unformattable
```
Do you want to add javadocs (i.e., what does it mean to be "unformattable")?

jon-wei commented on core/src/main/java/org/apache/druid/data/input/InputSourceReader.java:
```diff
@@ -0,0 +1,42 @@
+ * InputSourceReader reads data from {@link InputSource} and returns a {@link CloseableIterator} of
+ * {@link InputRow}. See {@link InputSource} for an example usage.
+ *
+ * Implementations of this class can use {@link ObjectSource} and {@link ObjectReader}.
```
Suggest linking to InputFormat along with ObjectSource and ObjectReader since the InputFormat would be used to create the ObjectReader if an impl chooses to use these classes

jihoonson replied:
Added.

jon-wei commented on core/src/main/java/org/apache/druid/data/input/ObjectSource.java:
```diff
@@ -0,0 +1,116 @@
+   *
+   * @see #fetch as an alternative way to read data.
+   */
+  InputStream open() throws IOException;
```
For something that's planning to get data from an ObjectSource, how should it determine when to call `open()` vs `fetch()`?

jon-wei commented on core/src/main/java/org/apache/druid/data/input/ObjectSource.java:
```diff
@@ -0,0 +1,116 @@
+ * ObjectSource abstracts an object and knows how to read bytes from the given object.
+ */
+@ExtensionPoint
+public interface ObjectSource<T>
```
Suggest putting "Input" or "Ingestion" in the name of this interface so that it's clear that it's used for ingestion (maybe `InputEntity` and `InputEntityReader`)?

jihoonson replied:
👍

jihoonson commented on core/src/main/java/org/apache/druid/data/input/FiniteFirehoseFactory.java:
```diff
@@ -33,6 +33,7 @@
  * @param <T> parser type
  * @param <S> input split type
  */
+@Deprecated
```
👍 Added.

jihoonson commented on core/src/main/java/org/apache/druid/data/input/InputSource.java:
```diff
@@ -0,0 +1,78 @@
+  boolean isSplittable();
+
+  /**
+   * Returns true if this inputSource supports different {@link InputFormat}s.
```
Added. Didn't link to `druid inputSource` here yet since it will be implemented in a follow-up pr.

jihoonson commented on core/src/main/java/org/apache/druid/data/input/TextReader.java:
```diff
@@ -0,0 +1,134 @@
+  public abstract boolean needsToProcessHeaderLine();
+
+  /**
+   * Processes a header line. This will be called as many times as {@link #getNumHeaderLinesToSkip()}.
```
Oops 👍

jon-wei commented on indexing-service/src/main/java/org/apache/druid/indexing/common/task/batch/parallel/ParallelIndexPhaseRunner.java:
```diff
@@ -255,15 +261,16 @@ public void onFailure(Throwable t)
-    if (firehoseFactory instanceof FiniteFirehoseFactory) {
-      final FiniteFirehoseFactory<?, ?> finiteFirehoseFactory = (FiniteFirehoseFactory) firehoseFactory;
-      return finiteFirehoseFactory.getSplits(splitHintSpec).collect(Collectors.toList());
+    if (inputSource instanceof SplittableInputSource) {
```
This could call `inputSource.isSplittable()` instead

jon-wei commented on core/src/test/java/org/apache/druid/data/input/impl/NoopFirehoseFactory.java:
```diff
@@ -0,0 +1,54 @@
+import javax.annotation.Nullable;
+import java.util.stream.Stream;
+
+public class NoopFirehoseFactory implements FiniteFirehoseFactory
```
Suggest calling these classes `NoopTest*`

jihoonson replied:
Hmm, I will keep this for now since all our noop implementations for tests start with `Noop` instead of `NoopTest`. Probably will do it in a separate pr if it's necessary.

clintropolis commented on core/src/main/java/org/apache/druid/data/input/TextReader.java:
```diff
@@ -0,0 +1,134 @@
+  /**
+   * Parses the given line into {@link InputRow}.
+   */
+  public abstract InputRow readLine(String line) throws IOException, ParseException;
```
I think this should be future proofed to return a list of `InputRow` (for explode spec)

jihoonson replied:
👍

clintropolis commented on core/src/main/java/org/apache/druid/data/input/InputSourceReader.java:
```diff
@@ -0,0 +1,43 @@
+{
+  CloseableIterator<InputRow> read() throws IOException;
+
+  CloseableIterator<InputRowPlusRaw> sample() throws IOException;
```
To handle potential 1:many mapping of raw input records to `InputRow`, we probably want a new wrapper type that associates a `List<InputRow>` to the 'raw' data

jihoonson replied:
👍

clintropolis commented on core/src/main/java/org/apache/druid/data/input/AbstractInputSource.java:
```diff
@@ -0,0 +1,65 @@
+    throw new UnsupportedOperationException("Implement this method properly if needsFormat() = true");
+  }
+
+  protected InputSourceReader unformattableReader(InputRowSchema inputRowSchema, @Nullable File temporaryDirectory)
```
i think `implicitlyFormattedReader` or `fixedFormatReader` might be a more intuitive name for this

jihoonson replied:
Changed to `fixedFormatReader`.

clintropolis commented on core/src/main/java/org/apache/druid/data/input/InputEntity.java:
```diff
@@ -0,0 +1,113 @@
+/**
+ * InputEntity abstracts an object and knows how to read bytes from the given object.
+ */
+@ExtensionPoint
```
Should we maybe hold off on adding `@ExtensionPoint` on the initial release of this to allow time for the interfaces to settle?

jihoonson replied:
Added `@UnstableApi` and changed with it. Same for other interfaces.

clintropolis commented on core/src/main/java/org/apache/druid/data/input/InputEntityReader.java:
```diff
@@ -0,0 +1,40 @@
+ *
+ * @see TextReader for text format readers
+ */
+@ExtensionPoint
```
same comment re `@ExtensionPoint`

clintropolis commented on core/src/main/java/org/apache/druid/data/input/InputFormat.java:
```diff
@@ -0,0 +1,57 @@
+ *
+ * @see NestedInputFormat for nested input formats such as JSON.
+ */
+@ExtensionPoint
```
same comment re `@ExtensionPoint`

clintropolis commented on core/src/main/java/org/apache/druid/data/input/InputSource.java:
```diff
@@ -0,0 +1,81 @@
+ *   }
+ * }</pre>
+ */
+@ExtensionPoint
```
same comment re `@ExtensionPoint`

clintropolis commented on core/src/main/java/org/apache/druid/data/input/InputSourceReader.java:
```diff
@@ -0,0 +1,43 @@
+ *
+ * See {@link InputEntityIteratingReader} as an example.
+ */
+@ExtensionPoint
```
same comment re `@ExtensionPoint`

jihoonson commented:
Updated PR to support the sampler better. Also addressed comments and merged master. Moved all tests in `CSVParserTest` into `CsvReaderTest`. `InputRowListPlusJson` has a list of `InputRow`s, a JSON string, and a byte array for now. Will clean up in my follow-up pr to make sampler working with the new interfaces.

clintropolis commented on core/src/main/java/org/apache/druid/data/input/impl/CsvReader.java:
```diff
@@ -0,0 +1,132 @@
+  {
+    final Map<String, Object> zipped = parseLine(line);
+    return Collections.singletonList(
+        MapInputRowParser.parse(
```
Can you add a version of `MapInputRowParser.parse` that takes an `InputRowSchema` since providing timestamp and dimension specs from it seems like it will be common

jihoonson replied:
Added.

clintropolis commented on processing/src/main/java/org/apache/druid/segment/transform/Transformer.java:
```diff
@@ -93,6 +93,31 @@ public InputRow transform(@Nullable final InputRow row)
+    if (transforms.isEmpty()) {
+      transformedRow = row;
+    } else {
+      transformedRow = InputRowListPlusJson.of(new TransformedInputRow(row.getInputRow(), transforms), row.getRaw());
```
should this be transforming the list of `InputRow` to `TransformedInputRow`?

jihoonson replied:
👍 fixed.

clintropolis commented on core/src/main/java/org/apache/druid/guice/annotations/UnstableApi.java:
```diff
@@ -0,0 +1,41 @@
+import java.lang.annotation.Target;
+
+/**
+ * Signifies that the annotated entity is an unstable API for extension authors. Unstable APIs may change at any time
```
this should also maybe indicate that someday it will likely become either an `@ExtensionPoint` or `@PublicApi`?

jihoonson replied:
👍 added.

jihoonson commented on core/src/main/java/org/apache/druid/data/input/InputEntity.java:
```diff
@@ -0,0 +1,115 @@
+   */
+  default CleanableFile fetch(File temporaryDirectory, byte[] fetchBuffer) throws IOException
+  {
+    final File tempFile = File.createTempFile("druid-object-source", ".tmp", temporaryDirectory);
```
👍

jihoonson commented on core/src/main/java/org/apache/druid/data/input/InputEntity.java:
```diff
@@ -0,0 +1,115 @@
+    File file();
+  }
+
+  T getObject();
```
Removed this and added `getUri()` instead.

jihoonson commented on core/src/main/java/org/apache/druid/data/input/Firehose.java:
```diff
@@ -74,13 +74,13 @@
    *
    * @return an InputRowPlusRaw which may contain any of: an InputRow, the raw data, or a ParseException
```
This method is only for sampler and will be removed in the follow-up pr.

jihoonson commented on core/src/main/java/org/apache/druid/data/input/InputRowListPlusJson.java:
```diff
@@ -62,23 +95,19 @@ public InputRow getInputRow()
+  public boolean isEmpty()
   {
-    return new InputRowPlusRaw(null, raw, parseException);
+    return (inputRows == null || inputRows.isEmpty()) && raw == null && rawJson == null && parseException == null;
```
This class is also used only by sampler and will be cleaned up in the follow-up pr.

jihoonson commented:
@ccaominh @clintropolis @jon-wei thanks for the review!
